### PR TITLE
Make Java code return reason for follow request failure

### PR DESCRIPTION
**Pull request by kevodwyer**

### CODE REVIEW

File: src/peergos/shared/user/UserContext.java [hunk 1]
```diff
@@ -799,21 +801,27 @@ private static void ensureAllowedToShare(FileWrapper file, String ourname, boole
...
             return sharing.getChildren(network).thenCompose(children -> {
+                CompletableFuture<Boolean> res = new CompletableFuture<>();
                 boolean alreadySentRequest = children.stream()
                         .filter(f -> f.getFileProperties().name.equals(targetUsername))
                         .findAny()
                         .isPresent();
-                if (alreadySentRequest)
-                    return CompletableFuture.completedFuture(false);
+                if (alreadySentRequest) {
+                    res.completeExceptionally(new Exception("Follow Request already sent!"));
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: Can use Futures.errored

File: src/peergos/shared/corenode/TofuCoreNode.java [hunk 2]
```diff
@@ -71,9 +71,17 @@ private static String getStorePath(String username) {
         if (local.isPresent())
             return CompletableFuture.completedFuture(local);
         return source.getChain(username)
-                .thenCompose(chain -> tofu.updateChain(username, chain, context.network.dhtClient)
-                        .thenCompose(x -> commit()))
-                .thenApply(x -> tofu.getPublicKey(username));
+                .thenCompose(chain -> {
+                        if(chain.isEmpty()) {
+                            CompletableFuture<Boolean> noSuchUser = new CompletableFuture<>();
+                            noSuchUser.complete(false);
```
Comment (ianopolous) [inline src/peergos/shared/corenode/TofuCoreNode.java]: Can use CompletableFuture.completedFuture

File: src/peergos/shared/user/UserContext.java [hunk 3]
```diff
@@ -601,8 +601,10 @@ public boolean isJavascript() {
     }
 
     public CompletableFuture<Optional<Pair<PublicKeyHash, PublicBoxingKey>>> getPublicKeys(String username) {
+        CompletableFuture<Optional<Pair<PublicKeyHash, PublicBoxingKey>>> empty = new CompletableFuture<>();
+        empty.complete(Optional.empty());
         return network.coreNode.getPublicKeyHash(username)
-                .thenCompose(signerOpt -> getSigningKey(signerOpt.get())
+                .thenCompose(signerOpt -> !signerOpt.isPresent() ? empty : getSigningKey(signerOpt.get())
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: Can use signingOpt.map(..).orElse(CompleteableFuture.completedFuture(Optional.empty()))

File: src/peergos/shared/user/UserContext.java (hunk 1 above)
Comment (kevodwyer) [reply src/peergos/shared/user/UserContext.java]: fixed

File: src/peergos/shared/user/UserContext.java (hunk 3 above)
Comment (kevodwyer) [reply src/peergos/shared/user/UserContext.java]: fixed

File: src/peergos/shared/corenode/TofuCoreNode.java (hunk 2 above)
Comment (kevodwyer) [reply src/peergos/shared/corenode/TofuCoreNode.java]: fixed

**Pull request merged**